error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MonoZipOfMonoVoidUsage.java

```diff
@@ -0,0 +1,143 @@
+ * cases this is not the desired behaviour.
+ *
+ * <p>NB: {@code Mono<?>.zipWith(Mono<Void>)} is allowed be the Reactor API, but it is an incorrect
+ * usage of the API. It will be flagged by ErrorProne but the fix won't be supplied. The problem
+ * with the original code should be revisited and fixed in a structural manner by the developer.
+ */
+@AutoService(BugChecker.class)
+@BugPattern(
+    summary =
+        "`Mono#zip` and `Mono#zipWith` should not be executed against `Mono#empty` or `Mono<Void>` parameter; "
+            + "please revisit the parameters used and make sure to supply correct publishers instead",
+    link = BUG_PATTERNS_BASE_URL + "MonoZipOfMonoVoidUsage",
+    linkType = CUSTOM,
+    severity = ERROR,
+    tags = LIKELY_ERROR)
+public final class MonoZipOfMonoVoidUsage extends BugChecker
+    implements MethodInvocationTreeMatcher {
+  private static final long serialVersionUID = 1L;
+  private static final Supplier<Type> MONO = type("reactor.core.publisher.Mono");
+  // In fact, we use `Mono<Void>` everywhere in codebases instead of `Mono<Object>` to represent
+  // empty publisher
+  private static final Supplier<Type> MONO_VOID_TYPE =
+      VisitorState.memoize(generic(MONO, type("java.lang.Void")));
+
+  // On instance mono.zipWith, at least one element should match empty in order to proceed.
+  private static final Matcher<ExpressionTree> MONO_ZIP_AND_WITH =
+      anyOf(
+          allOf(
+              instanceMethod().onDescendantOf(MONO).named("zipWith"),
+              toType(MethodInvocationTree.class, hasGenericArgumentOfExactType(MONO_VOID_TYPE))),
+          allOf(
+              instanceMethod().onDescendantOf(MONO).named("zipWith"),
+              toType(MethodInvocationTree.class, staticMethod().onClass(MONO).named("empty"))),
+          onClassWithMethodName(MONO_VOID_TYPE, "zipWith"));
+
+  // On class Mono.zip, at least one element should match empty in order to proceed.
+  private static final Matcher<ExpressionTree> STATIC_MONO_ZIP =
+      allOf(
+          staticMethod().onClass(MONO).named("zip"),
+          toType(MethodInvocationTree.class, hasGenericArgumentOfExactType(MONO_VOID_TYPE)));
+
+  /** Instantiates a new {@link MonoZipOfMonoVoidUsage} instance. */
+  public MonoZipOfMonoVoidUsage() {}
+
+  @Override
+  public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
+    boolean dynamicMono = MONO_ZIP_AND_WITH.matches(tree, state);
+    boolean staticMono = STATIC_MONO_ZIP.matches(tree, state);
+    if (!dynamicMono && !staticMono) {
+      return Description.NO_MATCH;
+    }
+
+    return buildDescription(tree)
+        .setMessage(
+            "`Mono#zip` and `Mono#zipWith` should not be executed against empty publisher; "
+                + "remove it or suppress this warning and add a comment explaining its purpose")
+        .addFix(SuggestedFixes.addSuppressWarnings(state, canonicalName()))
+        .build();
```

Comment:
I think the convention is to use a `describeMatch` method with a "built" `SuggestedFixBuilder`.

```suggestion
    return describeMatch(tree, fix);
```

Reply:
We don't need to use `setMessage` indeed, but @mohamedsamehsalah note the `SuggestedFix` provides a `@SuppressWarnings` :).